### PR TITLE
N + 1 prob milestone.

### PR DESCRIPTION
**Pull request by Nowembabazi**
#In this milestone i achieved the following;
-Write integration tests with Capybara gem.
-Secure app from n+1 problems.
Below are screenshots.
Before-
217188834-a56e0faa-27df-4d52-9c6b-efdcc138a415

After-
217188867-580f7200-fb86-4bea-9007-018be350b6ed

### CODE REVIEW

File: spec/views/posts/index.html.erb_spec.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      click_link('All comments')
+      expect(page).to have_current_path(user_post_path(@user, @first_post))
+    end
+
+    it 'redirects to posts show page' do
+      click_on 'All comments'
+      expect(page).to have_current_path user_post_path(@user.id, @post.id)
+    end
+  end
+end
```
Comment (KayLemba) [inline spec/views/posts/index.html.erb_spec.rb]: While you did a good job working on the project requirements whenever I run your tests there seems to be an error please kindly fix this issue.
Screenshot 2023-02-21 at 12 06 36 PM

**Review by KayLemba — CHANGES_REQUESTED** [review]
Hi @Nowembabazi    👋
Good job so far working on the project requirements!
There are some issues that you still need to work on to go to the next project but you are almost there!

To Highlight 🎉
No linter errors. ✔️
GitHub flow is followed. ✔️

Required Changes ♻️
Check the comments under the review in the inline comments.
 
Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

You can also consider:
N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.

As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Review by omar-muhamad — COMMENTED** [review]
# Invalid ❌

Hi @Nowembabazi,

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

Your Code Review Request is marked as invalid for these reasons:

- [ ] You used `sqlite3` for creating the database, Please use `postgres` instead to configure your database. 

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **Does not count** into the code reviews limit._

File: Gemfile [hunk 2]
```diff
@@ -9,8 +9,8 @@ gem 'rails', '~> 7.0.4', '>= 7.0.4.2'
 # The original asset pipeline for Rails [https://github.com/rails/sprockets-rails]
 gem 'sprockets-rails'
 
-# Use sqlite3 as the database for Active Record
-gem 'sqlite3', '~> 1.4'
+# Use postgresql as the database for Active Record
+gem 'pg', '~> 1.1'
```
Comment (1GORDON) [inline Gemfile:13]: Great job @Nowembabazi 🤝

- [ ] Just like the previous reviewer suggested, I keep getting this error whenever I try to run the tests.
error

**Review by 1GORDON — CHANGES_REQUESTED** [review]
##
## Hi @Nowembabazi ✋

Good job so far!

## Highlights!
- Followed Gitflow ✔️
- Linters are passing ✔️
- Descriptive Pull request ✔️
- Professional readme ✔️

Hello ✋, you have done well reaching this point of the milestone 🤝 . But there are some issues that you still need to work on to go to the next project but you are almost there!

# Status: Required Changes ♻️⛔️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread and tag me `@1GORDON` if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/post_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      ),
+      @user2 = User.create(
+        name: 'Lilly',
+        photo: 'https://img.freepik.com/free-vector/mysterious-mafia-man-smoking-cigarette_52683-34828.jpg?w=1480&t=st=1674661161~exp=1674661761~hmac=e529680000ea966150e5ea3b38241a0d8c9582faf23dac8af61ce1785cf27838',
+        bio: 'Teacher from Poland.', posts_counter: 1
+      )
+    ]
+
+    @posts = [
+      @post1 = Post.create(
```
Comment (Davidosky007) [inline spec/features/post_index_spec.rb:20]: * As the previous CR pointed out I still see some failing tests, Kindly fix them 🙏 
image

**Review by Davidosky007 — CHANGES_REQUESTED** [review]
Hi @Nowembabazi 👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
* No linter errors 👍 
* Well written PR description 👍 
* All tests passed ✔️ 
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Davidosky007 — APPROVED** [review]
Hi @Nowembabazi 👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
* No linter errors 👍 
* Well written PR description 👍 
* All tests passed ✔️ 
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**